File: Graph/Bellman-Ford/Bellman-Ford.py
class Graph:
    def __init__(self):
        self.vertices = set()
        self.edges = []

    def add_edge(self, start, end, weight):
        self.vertices.add(start)
        self.vertices.add(end)
        self.edges.append((start, end, weight))

    def bellman_ford(self, start):
        distances = {vertex: float('infinity') for vertex in self.vertices}
        distances[start] = 0

        for _ in range(len(self.vertices) - 1):
            for start_vertex, end_vertex, weight in self.edges:
                if distances[start_vertex] + weight < distances[end_vertex]:
                    distances[end_vertex] = distances[start_vertex] + weight

        # Проверка наличия отрицательных циклов
        for start_vertex, end_vertex, weight in self.edges:
            if distances[start_vertex] + weight < distances[end_vertex]:
                raise ValueError("Граф содержит отрицательный цикл")

        return distances


# Пример использования
graph = Graph()
graph.add_edge("A", "B", 1)
graph.add_edge("A", "C", 4)
graph.add_edge("B", "C", 2)
graph.add_edge("B", "D", 5)
graph.add_edge("C", "D", 1)
graph.add_edge("D", "A", -7)

start_vertex = "A"
result = graph.bellman_ford(start_vertex)
print(f"Shortest distances from vertex {start_vertex}: {result}")

File: Graph/Bellman-Ford/Bellman-Ford.cs
using System;
using System.Collections.Generic;

class Graph
{
    private HashSet<char> vertices = new HashSet<char>();
    private List<Tuple<char, char, int>> edges = new List<Tuple<char, char, int>>();

    public void AddEdge(char start, char end, int weight)
    {
        vertices.Add(start);
        vertices.Add(end);
        edges.Add(Tuple.Create(start, end, weight));
    }

    public Dictionary<char, int> BellmanFord(char start)
    {
        Dictionary<char, int> distances = new Dictionary<char, int>();
        foreach (char vertex in vertices)
        {
            distances[vertex] = int.MaxValue;
        }
        distances[start] = 0;

        for (int i = 0; i < vertices.Count - 1; ++i)
        {
            foreach (var edge in edges)
            {
                char startVertex = edge.Item1;
                char endVertex = edge.Item2;
                int weight = edge.Item3;

                if (distances[startVertex] + weight < distances[endVertex])
                {
                    distances[endVertex] = distances[startVertex] + weight;
                }
            }
        }

        // Проверка наличия отрицательных циклов
        foreach (var edge in edges)
        {
            char startVertex = edge.Item1;
            char endVertex = edge.Item2;
            int weight = edge.Item3;

            if (distances[startVertex] + weight < distances[endVertex])
            {
                throw new InvalidOperationException("Граф содержит отрицательный цикл");
            }
        }

        return distances;
    }
}

class Program
{
    static void Main()
    {
        Graph graph = new Graph();
        graph.AddEdge('A', 'B', 1);
        graph.AddEdge('A', 'C', 4);
        graph.AddEdge('B', 'C', 2);
        graph.AddEdge('B', 'D', 5);
        graph.AddEdge('C', 'D', 1);
        graph.AddEdge('D', 'A', -7);

        char startVertex = 'A';
        Dictionary<char, int> result = graph.BellmanFord(startVertex);

        Console.WriteLine($"Shortest distances from vertex {startVertex}:");
        foreach (var entry in result)
        {
            Console.WriteLine($"{entry.Key}: {entry.Value}");
        }
    }
}